### ISSUE TITLE
Effect manager review for mgold/elm-random-pcg

### ISSUE THREAD

**Issue by mgold** [elm-lang/package.elm-lang.org#168]
I'd like to allow randomness-as-a-service under the same API, and same effect implementation, as core/Random.

The primary API of `Random.Pcg` is that one can replace `import Random` with `import Random.Pcg as Random` and everything still works. Version 2.0.0 comes as close as possible without being an approved effects module. The [final difference](https://github.com/mgold/elm-random-pcg/commit/2c9c1e24e900f1237b0b6340ea3333a018eb4c40#diff-e582c4053a4e63d4879dcbd753ceec79) involves adding the `generate` function and all of the infrastructure that goes with it, again all copied straight from core.

**Comment by evancz** [comment]
Why not just call `Random.generate` from within your module? Define it in terms of the other one.

**Comment by mgold** [comment]
Because `Random.Generator` and `Random.Pcg.Generator` are two completely different types. They have to be so primitives like `int` and `float` can use a completely different RNG. Different algorithm, and different seed (it stores more bits).

`Random.generate` always returns a value computed with the core's RNG, whose [demonstrable poor statistical quality](https://github.com/mgold/elm-random-pcg/tree/master/test/dieharder#dieharder-tests-for-randompcg) is one of the reasons for Pcg's existence. The only way to build one on the other is to generate seeds from core's Random and feed them to Pcg:

```elm
int64 : Random.Generator
int64 =
  let
    int32 = Random.int 0 0xFFFFFFFF
  in
    Random.pair int32 int32

pcgGenerate tagger generator =
  let
    coreToPcg (seedA, seedB) =
      Pcg.step generator (Pcg.initialSeed2 seedA seedB) |> fst |> tagger
  in
    Random.generate coreToPcg int64
```
This is the same trick I use in `Random.Pcg.Interop` but even then it's not considered ideal and hopefully done only once. RNGs are garbage-in, garbage-out, and trying to generate seeds from an inferior source many times will leads to compromised statistical strength.

...I think. @jvoigtlaender am I misinformed about the statistical properties of [this approach](https://github.com/mgold/elm-random-pcg/commit/772e24bf35391b93370f989abac946608ca1998e) over many calls?

**Comment by jvoigtlaender** [comment]
I think Max is right.

**Comment by evancz** [comment]
If this implementation is better, why not just make it the default implementation? Why would we want the default to be bad, and now people need to know that and go to another library?

**Comment by mgold** [comment]
I largely agree with that reasoning, but two caveats:

1. "Please approve this effects module" is a smaller request than "Please merge this non-trivial PR into core".
2. Random.Pcg is pushing 1000 lines, although a good chunk of that is documentation and newlines. Still, that's a noticeable burden on maintenance and runtime size. Some of the extra size comes from an expanded API, which I think is also valuable, but merging into core could involve "do we really need function x" discussions. I'm the maintainer of elm-community/elm-random-extra, so I'd be able to move some functions into that. But that puts us back where we started, at why do I need this third-party library, why can't core do what I need?

If you're okay merging this fairly substantial contribution, I can draw up a PR to core.

**Comment by evancz** [comment]
The only thing that needs to change is how `Seed` works right? As long as you have `Random.andThen` you can create whatever API you want, so I do not expect there to be *any* API discussion.

**Comment by evancz** [comment]
Let me say it another way. There are two goals here.

  1. Get a "more random" implementation
  2. Expand the API to do lots of other things

We should solve number 1 now. As far as I can tell, this would result in a patch to core. No new functions, no API changes. It'd just change some backing code.

With number 2, nothing has changed about the advice of putting things in `*-extra` so that it can be considered later.

We should not mix these two separate concerns.

**Comment by mgold** [comment]
I'll add number 3, improve the documentation, as another orthogonal concern. What are your thoughts on merging the docs from my library?

> The only thing that needs to change is how `Seed` works right?

That, and the `int` and `float` functions on which everything else is built. But yes, I got you now, we can just change the implementation and keep the interface. 

You are correct that an opaque seed with `andThen` can support many more advanced APIs. But here are some that it can't:
* Splittable RNGs, implemented as `independentSeed`, which will be essential for the parallel testing I'm working on with Richard.
* JSON encoding/decoding of the seed, which was requested by someone using the library (can't find the link though)
* `Random.constant : a -> Generator a` can be implemented more efficiently within the module than by a third party. (This is m****ic `return`, i.e. simple but important.) 

All of these require expanding the API. I like the idea of merging the essential Pcg code into core and building up the API in -extra, but I consider splittable RNGs to block that (the other two are nice-to-have).

**Comment by evancz** [comment]
We figured out how to do this in person.
